### ISSUE TITLE
Implement run-time testing

### ISSUE THREAD

**Issue by t0mpr1c3** [AllYarnsAreBeautiful/ayab-desktop#413]
- [x]  Add machine states to desktop (#415)

- [x]  Add machine states to firmware run-time and transfer test functions from v0.80 standalone test build.

- [x]  Add [TEST] button to GUI. (#415)

- [x]  Implement serial monitor to display test information from device and elicit user input at prompts (e.g. https://doc.qt.io/qt-5/qtserialport-terminal-example.html) (#415)

- [x]  Document API changes  (https://github.com/AllYarnsAreBeautiful/ayab-manual/pull/13)

**Comment by chris007de** [comment]
Some thoughts on the "Test" button:
* What do you think about moving the "Test" button into the "Tools" section of the menu bar? I think/hope it is not that frequently used.
* If the button shall stay where it currently (b1252f6292d8a3ecbee0096206769c0419aa479b) is, then there is a race condition between the "Test" button and the enabled state of the port selection combo box. Currently, the "Tests" button is active after starting the software, but the combo box is not, because no pattern has been selected. So we might just enable the port selection UI elements on startup as well.

Test mode looks really nice, I am happy that this feature finally made it to the main software. Thanks a lot for that!
What do you think about removing the input text box and just use buttons (or a combo box and a send button) to trigger the various test functions? I could imagine that some users find it to be tedious to type in the commands manually.

**Comment by t0mpr1c3** [comment]
Thanks Chris, these are great suggestions! very constructive and helpful, and I think should lead to some real improvements.

> * What do you think about moving the "Test" button into the "Tools" section of the menu bar? I think/hope it is not that frequently used.

That's a good idea. I think perhaps on the Tools submenu?

> * If the button shall stay where it currently ([b1252f6](https://github.com/AllYarnsAreBeautiful/ayab-desktop/commit/b1252f6292d8a3ecbee0096206769c0419aa479b)) is, then there is a race condition between the "Test" button and the enabled state of the port selection combo box. Currently, the "Tests" button is active after starting the software, but the combo box is not, because no pattern has been selected. So we might just enable the port selection UI elements on startup as well.

Yes, and actually that is easily done because they are not part of the DockWidget UI.

> What do you think about removing the input text box and just use buttons (or a combo box and a send button) to trigger the various test functions? I could imagine that some users find it to be tedious to type in the commands manually.

That makes sense. And if we are going that route we might as well hard-code the test commands into the API, which would simplify the firmware considerably.

The API changes are the most fundamental, so that is the thing that would need to be done first.

**Comment by t0mpr1c3** [comment]
Draft API https://github.com/AllYarnsAreBeautiful/ayab-manual/blob/f26029779981bd2f4ff240b4e39b870ed7354469/docs/APIv6.rst#id3
